Make it possible to report GraphicsContext::save() call stack when restore() mismatch happens

shallawa opened the pull request:
#### af40a5372c47ec53b8b0223e96efd92d85aa51aa
<pre>
Make it possible to report GraphicsContext::save() call stack when restore() mismatch happens
<a href="https://bugs.webkit.org/show_bug.cgi?id=269216">https://bugs.webkit.org/show_bug.cgi?id=269216</a>
<a href="https://rdar.apple.com/122819202">rdar://122819202</a>

Reviewed by NOBODY (OOPS!).

Not only GraphicsContext::save/restore have to balanced, but also their purposes
have to match. After 267803@main, we assert if the restore purpose does not match
the purpose of the current GraphicsContext state.

Although this assertion is helpful for debugging, sometimes it is hard to find out
where the save() call was issued.

To fix this, the save() call stack will be stored for debugging only and will be
printed before the restore() mismatch assertion. So we will have the unrestored
save() and the mismatch restore() call stacks.

* Source/WTF/wtf/PlatformEnable.h:
* Source/WebCore/platform/graphics/GraphicsContext.cpp:
(WebCore::GraphicsContext::save):
(WebCore::GraphicsContext::restore):
* Source/WebCore/platform/graphics/GraphicsContext.h:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/af40a5372c47ec53b8b0223e96efd92d85aa51aa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39499 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18478 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41853 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42033 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/35399 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/41805 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21359 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15807 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33009 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40073 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15568 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34208 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13522 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/13495 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35157 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43311 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/32952 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35860 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35488 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39286 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/39125 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/14311 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11797 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/37546 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/15917 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/46131 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/15965 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/9404 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/15574 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ebf75a9f29189b582ecf1ff2a38afe9ef63d7704. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/24267

shallawa committed:
Make it possible to report GraphicsContext::save() call stack when restore() mismatch happens
https://bugs.webkit.org/show_bug.cgi?id=269216
rdar://122819202

Reviewed by NOBODY (OOPS!).

Not only GraphicsContext::save/restore have to balanced, but also their purposes
have to match. After 267803@main, we assert if the restore purpose does not match
the purpose of the current GraphicsContext state.

Although this assertion is helpful for debugging, sometimes it is hard to find out
where the save() call was issued.

To fix this, the save() call stack will be stored for debugging only and will be
printed before the restore() mismatch assertion. So we will have the unrestored
save() and the mismatch restore() call stacks.

* Source/WTF/wtf/PlatformEnable.h:
* Source/WebCore/platform/graphics/GraphicsContext.cpp:
(WebCore::GraphicsContext::save):
(WebCore::GraphicsContext::restore):
* Source/WebCore/platform/graphics/GraphicsContext.h:

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/12f8e6132c140ac3a1c7da298401e8d05ade695f. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/24267

kkinnunen-apple commented on Source/WebCore/platform/graphics/GraphicsContext.h:
```diff
@@ -391,6 +392,10 @@ class GraphicsContext {
     bool m_contentfulPaintDetected : 1 { false };
+
+#if ENABLE(CONTEXT_SAVE_STACK_CAPTURE)
+    Vector<std::unique_ptr<WTF::StackShot>, 1> m_saveStacks;
```
StackShot should support automatic allocation, so you might be able to do Vector<WTF::StackShot>

kkinnunen-apple reviewed (approved):
Just some thoughts, not neccessarily anything needing action:
- How is it possible to have a mismatch? Intuitively all WebKit callsites should use RAII scoped holder to save and restore, as there should never be a active save that escapes the call frame.. The quickly thinking the only case we need to support escaping save() is 2D context. All this could be guaranteed by C++ type system by making save()/restore() private and the RAII context saver a friend.
- It'd be great to have default maxFramesToCapture = 0  to turn the feature off and not have the ifdef, since off-by-default ifdefs bitrot quite fast. I don't have good suggestions how to do it, though, since it'd be bad to pay for unused stackshot member variable.
- It would appear that the m_saveStacks contains the exact same amount of entries as m_stack. It could be thus possible to put the stack shot into GraphicsContextState, thus automatically keeping the m_state and the stack shot association more reliable as they never can get disassociated.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/af40a5372c47ec53b8b0223e96efd92d85aa51aa. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/24267

shallawa commented:
Since the mismatched save()/restore() purpose is not very common, I will convert this PR to draft. This change does not need to be in the build. When needed only, the PR can be checkered out for debugging.
